PROJ-101 - Sliding window protocol (#56)
SW805F20/report: sections/sprint-3/sliding-window-protocol.tex

\section{Sliding window protocol}
In \autoref{subsec:data-format} it was decided that the format of timestamps would be an integer from 0-255, where 0 to 20 would be considered as newer timestamps than 235 to 255.

> However, if all the timestamps between 0 to 20, there would not be any updates until it has looped to a number between 0 and 20 and sent new updates.

Reviewer: Forstår ikke rigtig hvad du prøver at sige på den her linje

> However, if all the timestamps between 0 to 20, there would not be any updates until it has looped to a number between 0 and 20 and sent new updates.

Reviewer 2: ```suggestion
However, if all the timestamps from 0 to 20 were lost, there would not be any updates until it has iterated through all other timestamps and back to 0.
```

> To find a better solution to sliding window protocol was researched.

Reviewer 2: ```suggestion
To find a better solution the sliding window protocol was researched.
```

> To find a better solution to sliding window protocol was researched.

Reviewer: ```suggestion
To find a better solution to this problem sliding window protocol was researched.
```

\\\\

> Sliding window protocols are often used where reliable data is required.

Reviewer: Hvad er reliable data ? ud dyb det mere. Det her hænger sammen med min næste kommentar

> Sliding window protocols are often used where reliable data is required.

Reviewer: hvad betyder reliable data?

> Whenever the senders has sent the window sized packets, it waits for acknowledgement from the receiver before it sends new packets.

Reviewer: Introen i sektionen generelt er lidt tynd. Overvej at forklare lidt mere om hvad problemet er, hvad en sliding protocol og hvorfor den løser hvores problem

> Whenever the senders has sent the window sized packets, it waits for acknowledgement from the receiver before it sends new packets.

Reviewer: hvad er en window sized packet ?

> Whenever the senders has sent the window sized packets, it waits for acknowledgement from the receiver before it sends new packets.

Reviewer 2: ```suggestion
Whenever the sender has sent the window sized packets, it waits for acknowledgement from the receiver before it sends new packets.
```

> Whenever the sender has sent the window sized packets, it waits for acknowledgement from the receiver before it sends new packets.

Reviewer: hvad er en window sized packet ?

> Each packet gets a sequence number and the acknowledgement sends back this number.

Reviewer: ```suggestion
Each packet gets a sequence number and the acknowledgement that is sent back also contains the sequence number.
```

By doing this the receiver can keep track of the packets, remove duplicates, identify missing packets and know which order is the correct one.
\\\\

> On \autoref{fig:sliding-window} there is an illustration of an ideal sliding window, where W is the window size, and n is the number of packets that has not been acknowledged.

Reviewer: ```suggestion
On \autoref{fig:sliding-window} there is an illustration of an ideal sliding window, where W is the window size, and n is the number of packets that have not been acknowledged.
```

> The sender will send the packets until the window is full and then wait for an acknowledgement from the sent packets.

Reviewer: ```suggestion
The sender will send the packets until the window is full and then wait for acknowledgements from the receiver for the sent packets.
```

\begin{figure}[H]
    \centering
    \includegraphics[width=0.6\linewidth]{window-protocol-illustration.PNG}
    \caption{Sliding window protocol without timeouts or message loss \cite{design-and-validation-of-computer-protocols}.}
    \label{fig:sliding-window}
\end{figure}
\noindent

> For our project reliable data in the format of goal zones and goal score is important, however the player position is not important that it is reliable.

Reviewer 2: ```suggestion
For our project reliable data in the format of goal zones and goal score is important, however the reliability of player positions are not important.
```

> This is due to our prioritization that the update rate is higher and hereby we previously choose to use UDP in \autoref{subsubsec:choosing-between-udp-or-tcp}.

Reviewer 2: hva? måske ikke nødvendig eller så forklar lidt bedre hvorfor udp og update rate er prioriteret bedre eller bare skrive onget i stil af " This is explain why in /sec bla bla"

\section{Sliding window protocol}
In \autoref{subsec:data-format} it was decided that the format of timestamps would be an integer from 0-255, where 0 to 20 would be considered as newer timestamps than 235 to 255.
However, if all the timestamps from 0 to 20 were lost, there would not be any updates until it has iterated through all other timestamps and back to 0.
To find a better solution to this problem sliding window protocol was researched.
\\\\
Sliding window protocols are often used where reliable data is required, which means making sure that all data points are received.

> This protocol uses a window sized packet, which is a fixed number where it sends this number of the elements to the receiver.

Reviewer: ```suggestion
This protocol uses a window sized packet, which is a fixed number where it sends this number of packages to the receiver.
```

\section{Sliding window protocol}
In \autoref{subsec:data-format} it was decided that the format of timestamps would be an integer from 0-255, where 0 to 20 would be considered as newer timestamps than 235 to 255.
However, if all the timestamps from 0 to 20 were lost, there would not be any updates until it has iterated through all other timestamps and back to 0.

> To find a better solution to this problem sliding window protocol was researched.

Reviewer 2: ```suggestion
To find a better solution to this problem the sliding window protocol was researched.
```

\\\\
Sliding window protocols are often used where reliable data is required, which means making sure that all data points are received.

> This protocol uses window sized buffer space, which is a fixed number that sends this number of the packets to the receiver \cite{design-and-validation-of-computer-protocols}.

Reviewer 2: ```suggestion
This protocol uses window sized buffer space, where the window is a fixed number and the sender will send this number of the packets to the receiver \cite{design-and-validation-of-computer-protocols}.
```

> Whenever the sender has sent the window sized packets, it waits for acknowledgement from the receiver before it sends new packets.

Reviewer: ```suggestion
Whenever the sender has sent the same number of packets as the size of the window, it waits for acknowledgements from the receiver before sending additional packets.
```

> Whenever the sender has sent the same number of packets as the size of the window, it waits for acknowledgements from the receiver before sending additional packets.

Reviewer 2: ```suggestion
Whenever the sender has sent the same number of packets as the size of the window, it waits for acknowledgements that the receiver has received the packets before sending additional packets.
```

Each packet gets a sequence number and the acknowledgement that is sent back also contains the sequence number.
By doing this the receiver can keep track of the packets, remove duplicates, identify missing packets and know which order is the correct one.
\\\\
On \autoref{fig:sliding-window} there is an illustration of a sliding window, where the window size is three.

> In the first frame three packets are send to the receiver, and in the second frame the receiver have acknowledged all three packets, but the first acknowledgement fails.

Reviewer: ```suggestion
In the first frame three packets are sent to the receiver, in the second frame the receiver has acknowledged all three packets, but the first acknowledgement fails.
```

> The sliding window for the receiver has moved its window, but as the sender only has an acknowledgement from the second and third packet, it is unable to move its window, and resend the packet in the third frame.

Reviewer: ```suggestion
The sliding window for the receiver has moved its window, but as the sender only has recieved acknowledgements from the second and third packet, it is unable to move its window and instead sends the first packet again in the third frame.
```

> In the fourth frame the first packet has been acknowledged, and hereby the sliding window has moved for the sender and is able to send new packets.

Reviewer: ```suggestion
In the fourth frame the first packet has been acknowledged, and therefore the sliding window is moved for the sender and the sender is now able to send new packets.
```

\begin{figure}[H]
    \centering
    \includegraphics[width=\linewidth]{sliding-window-illustration.png}
    \caption{Sliding window protocol illustration where the top is the sender and the bottom is the receiver.}
    \label{fig:sliding-window}
\end{figure}
\noindent
For our project reliable data in the format of goal zones and goal score is important, however the reliability of player positions are not important.
The reasoning for this is explained in \autoref{subsubsec:choosing-between-udp-or-tcp}.

> With Sliding Window protocol a player position would be retransmitted if it previously failed, and these positions quickly become outdated.

Reviewer: ```suggestion
With Sliding Window protocol a player position would be retransmitted if it previously failed, and this position would now be outdated because the player might have moved since then.
```
